Sources/GRPC/GRPCClientStateMachine.swift

```diff
@@ -0,0 +1,588 @@
+           (.clientClosedServerStreaming, .clientClosedServerClosed):
+        break
+
+      // Self transitions, also valid:
+      case (.clientIdleServerIdle, .clientIdleServerIdle),
+           (.clientStreamingServerIdle, .clientStreamingServerIdle),
+           (.clientClosedServerIdle, .clientClosedServerIdle),
+           (.clientStreamingServerStreaming, .clientStreamingServerStreaming),
+           (.clientClosedServerStreaming, .clientClosedServerStreaming),
+           (.clientClosedServerClosed, .clientClosedServerClosed):
+        break
+
+      default:
+        preconditionFailure("invalid state transition from '\(self.state)' to '\(oldValue)'")
+      }
+    }
+  }
+
+  private let logger: Logger
+
+  /// Creates a state machine representing a gRPC client's request and response stream state.
+  ///
+  /// - Parameter requestCount: The arity of the request stream.
+  /// - Parameter responseCount: The arity of the response stream.
+  /// - Parameter logger: Logger.
+  init(
+    requestCount: MessageCount,
+    responseCount: MessageCount,
+    logger: Logger
+  ) {
+    let pendingWriteState = PendingWriteState(
+      expectedCount: requestCount,
+      encoding: .none,
+      contentType: .protobuf
+    )
+    self.state = .clientIdleServerIdle(pendingWriteState, responseCount)
+    self.logger = logger
+  }
+
+  /// Creates a state machine representing a gRPC client's request and response stream state.
+  ///
+  /// - Parameter state: The initial state of the state machine.
+  /// - Parameter logger: Logger.
+  init(
+    state: State,
+    logger: Logger
+  ) {
+    self.state = state
+    self.logger = logger
+  }
+
+  /// Initiates an RPC.
+  ///
+  /// The only valid state transition is:
+  /// - `.clientIdleServerIdle` → `.clientStreamingServerIdle`
+  ///
+  /// All other states will result in a `.fatal` error.
+  ///
+  /// On success the state will transition to `.clientStreamingServerIdle`.
+  ///
+  /// - Parameter host: The host which will handle the RPC.
+  /// - Parameter path: The path of the RPC (e.g. '/echo.Echo/Collect').
+  /// - Parameter options: Options for this RPC.
+  /// - Parameter requestID: The uniuqe ID of this request used for logging.
+  mutating func sendRequestHeaders(
+    host: String,
+    path: String,
+    options: CallOptions,
+    requestID: String
+  ) -> Result<HTTPRequestHead, InvalidStateError> {
+    return self.state.sendRequestHeaders(
+      host: host,
+      path: path,
+      options: options,
+      requestID: requestID
+    )
+  }
+
+  /// Formats a request to send to the server.
+  ///
+  /// The client must be streaming in order for this to return successfully. Therefore the valid
+  /// state transitions are:
+  /// - `.clientStreamingServerIdle` → `.clientStreamingServerIdle`
+  /// - `.clientStreamingServerStreaming` → `.clientStreamingServerStreaming`
+  ///
+  /// It is invalid (but not fatal) for the client to attempt to send requests once the request
```

Comment:
Could we define "invalid" and "fatal" somewhere?

Reply:
Ah, the previous iteration of this used `.nonFatal` and `.fatal` errors to determine whether the RPC should be terminated or not: I'll do a pass of the documentation to update this.